### DISCUSSION TITLE
request  AbortController

### DISCUSSION THREAD

**Discussion by zj91zj** [umijs/umi#11963] [Q&A]
import { request } from '@umijs/max';
import { Button } from 'antd';

let controller;
const Page = () => {
  const handleSubmit = () => {

    controller = new AbortController();
    const { signal } = controller;
    request('/doLogin', { method: 'post', data: { username: 'admin', password: '123456' }, signal })
    setTimeout(() => {
      controller.abort();
    }, 1)
  }

  return (
    <Button onClick={() => { handleSubmit() }}>请求取消</Button>
  );
};
export default Page;

连接项目mock数据的时候，浏览器F12的网络请求中，1ms后能看到该条请求状态代码:已取消;
但是连接到正式环境api后，浏览器F12的网络请求中，该条请求状态代码:200 OK，该条请求还是在请求，只不过没有结果返回

**Comment by zj91zj** [comment]
我最终的用途是页面每隔1s会发起10个请求，直接取消上一次的10个请求。 但是使用controller.abort()后，页面依然堆积很多请求队列，有没有大佬能解答一下，🙇‍

**Comment by fz6m** [comment]
我不知道自带的 request 是否支持 signal ，因为这个功能几乎 99% 的人不会使用。

不推荐继续使用自带的 request 了，因为这是 umi 3 残留的兼容，现代数据流更推荐自行编写 axios / fetch 来发送请求，顶层使用 [swr](https://github.com/vercel/swr) 和 [react-query](https://umijs.org/docs/max/react-query) 管理请求状态，其中在自行编写 axios / fetch 的底层中，你可以通过对应的配置来管理请求的取消。

如果只是为了请求限制，请对发送请求的来源添加 [防抖](https://www.lodashjs.com/docs/lodash.debounce) 或节流 ，通过限制请求数而不是取消请求的方式进行。

**Comment by zj91zj** [comment]
安装这个可以取消

npm install --save yet-another-abortcontroller-polyfill

安装完成后，在代码中引入
复制代码import 'yet-another-abortcontroller-polyfill'

假如你是用umi-request做网络请求，使用方式如下：
复制代码import request from 'umi-request'

const controller = new AbortController()
const { signal } = controller

request('/api/path', {
    signal // 这个标记将信号和控制器关联，允许在fetch请求中中止请求
})

// 当你需要取消请求的时候,执行：
controller.abort()
